bench/wtperf/wtperf_throttle.c

```diff
@@ -0,0 +1,117 @@
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "wtperf.h"
+
+/*
+ * Put the initial config together for running a throttled workload.
+ */
+int
+setup_throttle(CONFIG_THREAD *thread)
+{
+	THROTTLE_CONFIG *throttle_cfg;
+
+	throttle_cfg = &thread->throttle_cfg;
+
+	/*
+	 * Setup how the number of operations to run each interval in order to
+	 * meet our desired max throughput.
+	 * - If we have a very small number of then we can look to do one op
+	 *   on a larger increment. Given there is overhead in throttle logic
+	 *   we want to avoid running the throttle check regularly.
+	 * - For most workloads, we aim to do 100 ops per interval and adjust
+	 *   the sleep period accordingly.
+	 * - For high throughput workloads, we aim to do many ops in 100us
+	 *   increments.
+	 */
+
+	if (thread->workload->throttle < THROTTLE_OPS) {
+		/* If the interval is very small, we do laps of 1 */
+		throttle_cfg->usecs_increment =
+		    USEC_PER_SEC / thread->workload->throttle;
+		throttle_cfg->ops_per_increment = 1;
+	} else if (thread->workload->throttle < USEC_PER_SEC / THROTTLE_OPS) {
+		throttle_cfg->usecs_increment =
+		    USEC_PER_SEC / thread->workload->throttle * THROTTLE_OPS;
+		throttle_cfg->ops_per_increment = THROTTLE_OPS;
+	} else {
+		/* If the interval is large, we do more ops per interval */
+		throttle_cfg->usecs_increment = USEC_PER_SEC / THROTTLE_OPS;
+		throttle_cfg->ops_per_increment =
+		    thread->workload->throttle / THROTTLE_OPS;
+	}
+
+	/* Give the queue some initial tickets to work with */
+	throttle_cfg->ops_count = throttle_cfg->ops_per_increment;
+
+	/* Set the first timestamp of when we incremented */
+	WT_RET(__wt_epoch(NULL, &throttle_cfg->last_increment));
+	return (0);
+}
+
+/*
+ * Run the throttle function. Will sleep if needed and then reload the counter
+ * to perform more operations.
+ */
+int
+worker_throttle(CONFIG_THREAD *thread)
+{
+	THROTTLE_CONFIG *throttle_cfg;
+	struct timespec now;
+	uint64_t usecs_delta;
+
+	throttle_cfg = &thread->throttle_cfg;
+
+	WT_RET(__wt_epoch(NULL, &now));
+
+	/*
+	 * If we did enough operations in the current interval, sleep for
+	 * the rest of the interval. Then add more tickets to the queue.
+	 */
+	usecs_delta = WT_TIMEDIFF_US(now, throttle_cfg->last_increment);
+	if (usecs_delta < throttle_cfg->usecs_increment) {
+		(void)usleep(
+		    (useconds_t)(throttle_cfg->usecs_increment - usecs_delta));
+		throttle_cfg->ops_count =
+		     throttle_cfg->ops_per_increment;
+	} else
+		throttle_cfg->ops_count =
+		    (uint64_t) (float)(usecs_delta /
+		    throttle_cfg->usecs_increment) *
+		    (float)throttle_cfg->ops_per_increment;
+
+	/* Don't over-fill the queue */
+	throttle_cfg->ops_count =
+	    WT_MIN(throttle_cfg->ops_count, thread->workload->throttle);
+
+	/*
+	 * After sleeping, set the interval to the current time.
+	 */
+	WT_RET(__wt_epoch(NULL, &throttle_cfg->last_increment));
```

Comment:
Can we avoid the second call to `__wt_epoch` in the fast path (where we aren't sleeping) and just set `throttle_cfg->last_increment = now`?